sycl/test-e2e/ESIMD/ext_math.cpp

```diff
@@ -361,20 +361,25 @@ bool test(queue &Q, const std::string &Name,
   int ErrCnt = 0;
 
   for (unsigned I = 0; I < Size; ++I) {
-    T Gold;
+    // functions like std::isinf/isfinite/isnan do not work correctly with
+    // sycl::half, thus we'll use 'float' instead.
+    using CheckT = std::conditional_t<std::is_same_v<T, sycl::half>, float, T>;
+
+    CheckT Gold;
 
     if constexpr (IsBinOp) {
       Gold = HostFunc<T, Op>{}((T)A[I], (T)B[I]);
     } else {
       Gold = HostFunc<T, Op>{}((T)A[I]);
     }
-    T Test = C[I];
+    CheckT Test = C[I];
 
     if (delta == 0.0f) {
       delta = sizeof(T) > 2 ? 0.0001 : 0.01;
     }
 
-    if (abs(Test - Gold) > delta) {
+    bool BothFinite = std::isfinite(Test) && std::isfinite(Gold);
```

Comment:
A case where one value is nan, and other is finite, also should be signaled as error.

Reply:
I was going to do the same originally, but std::isnan and std::isinf should not be used with -ffast-math. 
The attempt to do so causes warnings and also potential return of replacement of isnan with 'false' even when the arg is nan.

Reply:
The numbers go to different fp class on host and device easily because the input values include all possible values and the test function is one for all functions (pow, log, cos, etc).
The only certainty is when both test and gold values are finite numbers.